pandas/core/reshape/merge.py

```diff
@@ -1757,13 +1770,15 @@ def flip(xs):
             return func(left_values, right_values, self.allow_exact_matches, tolerance)
 
 
-def _get_multiindex_indexer(join_keys, index, sort):
+def _get_multiindex_indexer(join_keys, index: MultiIndex, sort: bool):
 
     # bind `sort` argument
     fkeys = partial(_factorize_keys, sort=sort)
 
     # left & right join labels and num. of levels at each location
-    rcodes, lcodes, shape = map(list, zip(*map(fkeys, index.levels, join_keys)))
+    mapped = [fkeys(index.levels[n], join_keys[n]) for n in range(len(index.levels))]
+    zipped = zip(*mapped)
+    rcodes, lcodes, shape = [list(x) for x in zipped]
```

Comment:
was this cleaning or silencing mypy?

Reply:
this was necessary to silence mypy.  i also dont like the pattern that was used before (and am not wild about the `zip` still here, open to suggestions)

Reply:
yep. fine with this. just curious.

Reply:
OK with splitting this out but should probably keep `mapped` as a generator; coercing to list may have some overhead